### PR TITLE
Make _replace more lenient.

### LINKED ISSUE #5361: Inconsistent behavior in grouby depending on the dimension order
`groupby` works inconsistently depending on the dimension order of a `DataArray`.
Furthermore, in some cases, this causes a corrupted object. 

```python
In [4]: data = xr.DataArray(
   ...:     np.random.randn(4, 2),
   ...:     dims=['x', 'z'],
   ...:     coords={'x': ['a', 'b', 'a', 'c'], 'y': ('x', [0, 1, 0, 2])}
   ...: )
   ...: 
   ...: data.groupby('x').mean()
Out[4]: 
<xarray.DataArray (x: 3, z: 2)>
array([[ 0.95447186, -1.14467028],
       [ 0.76294958,  0.3751244 ],
       [-0.41030223, -1.35344548]])
Coordinates:
  * x        (x) object 'a' 'b' 'c'
Dimensions without coordinates: z
```
`groupby` works fine (although this drops nondimensional coordinate `y`, related to #3745). 

However, `groupby` does not give a correct result if we work on the second dimension,
```python
In [5]: data.T.groupby('x').mean()  # <--- change the dimension order, and do the same thing
Out[5]: 
<xarray.DataArray (z: 2, x: 3)>
array([[ 0.95447186,  0.76294958, -0.41030223],
       [-1.14467028,  0.3751244 , -1.35344548]])
Coordinates:
  * x        (x) object 'a' 'b' 'c'
    y        (x) int64 0 1 0 2  # <-- the size must be 3!!
Dimensions without coordinates: z
```

The bug has been discussed in #2944 and solved, but I found this is still there.

<details><summary>Output of <tt>xr.show_versions()</tt></summary>

INSTALLED VERSIONS
------------------
commit: 09d8a4a785fa6521314924fd785740f2d13fb8ee
python: 3.7.7 (default, Mar 23 2020, 22:36:06) 
[GCC 7.3.0]
python-bits: 64
OS: Linux
OS-release: 5.4.0-72-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.10.4
libnetcdf: 4.6.1

xarray: 0.16.1.dev30+g1d3dee08.d20200808
pandas: 1.1.3
numpy: 1.18.1
scipy: 1.5.2
netCDF4: 1.4.2
pydap: None
h5netcdf: 0.8.0
h5py: 2.10.0
Nio: None
zarr: None
cftime: 1.2.1
nc_time_axis: None
PseudoNetCDF: None
rasterio: None
cfgrib: None
iris: None
bottleneck: None
dask: 2.6.0
distributed: 2.7.0
matplotlib: 3.2.2
cartopy: None
seaborn: 0.10.1
numbagg: None
pint: None
setuptools: 46.1.1.post20200323
pip: 20.0.2
conda: None
pytest: 5.2.1
IPython: 7.13.0
sphinx: None

</details>

### PR DESCRIPTION
**Pull request by dcherian**
- [x] Closes #5361
- [x] Tests added

### CODE REVIEW

File: xarray/core/dataarray.py [hunk 1]
```diff
@@ -534,10 +534,14 @@ def _replace_maybe_drop_dims(
         variable: Variable,
         name: Hashable | None | Default = _default,
     ) -> Self:
-        if variable.dims == self.dims and variable.shape == self.shape:
+        dims_are_equal = set(variable.dims) == set(self.dims)
+        size_is_equal = (
```
Comment (dcherian) [inline xarray/core/dataarray.py:538]: I don't see why this function used strict equality before

File: xarray/core/dataarray.py [hunk 2]
```diff
@@ -534,10 +534,14 @@ def _replace_maybe_drop_dims(
...
+            self.sizes[dim] == size for dim, size in variable.sizes.items()
+        )
+        if dims_are_equal and all(size_is_equal):
```
Comment (max-sixty) [inline xarray/core/dataarray.py:541]:
```suggestion
        variable.sizes == dims.sizes
```
?

(Seems to not check order)

**Review by max-sixty — APPROVED** [review]
Nice, thank you!

File: xarray/core/dataarray.py (hunk 2 above)
Comment (dcherian) [reply xarray/core/dataarray.py:541]: 🤦🏾‍♂️

**Comment by dcherian** [comment]
flaky failure is unrelated.

**Pull request merged**